docs/design/2025-04-15-global-memory-arbitrator.md

```diff
@@ -0,0 +1,435 @@
+
+- **[opt 1]** Set soft-limit to `AUTO` and let mem-arbitrator adaptively adjust the upper limit of global memory resources and gradually converges the OOM problem (multiple OOMs may occur in extreme cases).
+- **[opt 2]** Set the soft-limit to a relative small value / percentage, manually set the global memory quota limit, and quickly solve OOM problems
+- **[opt 3]** For the known SQL with memory leak problem, make it pre-subscribe a relative large amount of mem quota.
+
+Ensure the successful execution of important SQL
+
+- Bind the SQL with mem-priority `HIGH` and ensure that the SQL execution won't be Kill/Cancel
+- Make the SQL pre-subscribe a sufficient / excessive amount of mem quota
+  - Avoid blocking overhead caused by multiple subscriptions during execution
+- Under `PRIORITY` mode of mem-arbitrator, SQL can be bound with `wait_averse` property first. If the upper-layer retries multiple times and fails (the overall cluster resources are insufficient), disable `wait_averse` and retry by mem-priority
+
+### Architecture
+
+![mem-arbitrator](./imgs/mem-arbitrator.png)
+
+#### Memory Resource Pool
+
+`mem-pool` is a thread-safe structure which manages a tree-like set of quota, which consists of several cores:
+
+- `budget`:
+  - capacity
+  - used
+  - upstream mem-pool
+- limit
+- allocated
+- `reserved-budget`
+- `actions` interface: `notification`, `out-of-cap`, `out-of-limit`
+
+The allocated quota of `mem-pool` should not exceed the used quota of its `budget + reserved-budget`. When the used quota is GT than the capacity, the pool could increase the budget by **2** ways:
+
+- request from upstream mem-pool
+- `out-of-cap` action
+  - e.g. the root pool can request from its `mem-arbitrator`
+
+#### Memory Resource Arbitrator
+
+There will be a unique global-arbitrator under each TiDB instance. The arbitrator uses the common TiDB property `server-limit` as the hard limit for physical / logical memory, and sets it as the soft limit of the golang runtime through `debug.SetMemoryLimit`. The arbitrator also use `95% * server-limit` as the default mode of soft-limit.
+
+The mem quota is separated into **4** parts: `allocated`, `available`, `buffer`, `out-of-control`.
+
+- `allocated`: the quota occupied by mem pools (root `mem-pool` or [await-free-pool](#await-free-pool))
+- `available`: the quota available for alloc
+- [buffer](#buffer): the quota reserved to solve resolving [deadlock](#avoid-deadlock) or memory leaks under `PRIORITY` mode
+- [out-of-control](#out-of-control): the unsafe / prohibited quota
+  - The `soft-limit` takes effect by modifying out-of-control
+
+The arbitrator itself is a purely logical abstraction. It provides an async working method which can be executed in an independent goroutine.
+
+The root pool is required to implement **3** interfaces before attaching to the arbitrator:
+
+- `Cancel`
+  - Stop & release mem quota
+  - Invoked when the mem quota is insufficient
+- `Kill`
```

Comment:
What's the different between `cancel` and `kill`?

Reply:
By design, the difference between `cancel` and `kill` for SQL execution is the error code / msg. They both used the killer of mem tracker to send signal.
`kill` will only be invoked when facing `OOM Risk`.